Update simpletdm.pwn

Bronzal opened the pull request.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openmultiplayer/example-gamemodes?pullRequest=7) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/openmultiplayer/example-gamemodes?pullRequest=7) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openmultiplayer/example-gamemodes?pullRequest=7) it.</sub>

Bronzal commented:
* Fixed team assigner. 
* Removed default interior pickups ( in LVPD interior, people had chance to escape on street )
* Added statistics during player in session
* Added rewards such as Killing spree bonus and PlayerScore bonuses.

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -90,35 +109,39 @@ public OnGameModeExit()
+                {
+                    SetPlayerTeam(i,_:BLUE_TEAM),TEAM_BLUE_PLAYERS+=2;
+                }
+                else SetPlayerTeam(i,_:RED_TEAM),TEAM_RED_PLAYERS+=2;
```
Whats going on here? Horrible to read. Why not just split them on their own lines?

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -144,14 +169,18 @@ public OnPlayerSpawn(playerid)
+        {
+          SetPlayerTeam(playerid,_:BLUE_TEAM),TEAM_BLUE_PLAYERS+=2;
+          }
+        else SetPlayerTeam(playerid,_:RED_TEAM),TEAM_RED_PLAYERS+=2;
```
Same as on line 138. Horrible to read. Why not just split?

Bronzal replied:
Just used to do likes this and thats why..

chromaerror replied:
Do you mind changing it so others can also make sense of the code more easily? It's open source after all, no reason to make it so ridiculously annoying to read.
Especially as the `TEAM_BLUE_PLAYERS+=2;` is in no way connected to the `SetPlayerTeam` function, it would be logical for them to be on their own lines to drive the distinction more clear.

Bronzal replied:
Yes I will, but I not now, I’m on phone rn.

Bronzal replied:
TEAM_BLUE_PLAYERS +2 is used because of I wanted to each player join other team when they connect.

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -144,14 +169,18 @@ public OnPlayerSpawn(playerid)
+        if(TEAM_BLUE_PLAYERS <= TEAM_RED_PLAYERS)
+        {
+          SetPlayerTeam(playerid,_:BLUE_TEAM),TEAM_BLUE_PLAYERS+=2;
+          }
```
Indentation

Bronzal replied:
Oops missed that one.

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -5,25 +5,42 @@
+#define     KILLING_SPREE_STARTING_SCORE (3)
+#define     KILLING_SPREE_BONUS_MONEY (10000)
+#define     KILLING_SPREE_BONUS_SCORE (5)
+#define     ARMOUR_INCREMENT_STARTING_SCORE (5)
```
Unused define

Bronzal replied:
The last one isn’t used, I wrote code for this but then deleted, idk why did I decide that but next time Im going to implement that feature too.

chromaerror replied:
I guess that could be okay. However, as there's changes to be done here anyway, might as well take that away from this iteration as it is not used in this iteration

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -5,25 +5,42 @@
     DIALOG_HELP,
-    DIALOG_SHOP
+    DIALOG_SHOP,
+    DIALOG_STATS
```
`DIALOG_SHOP` and `DIALOG_STATS` are both unused

Bronzal replied:
Shop is used, same on stats - removed

chromaerror replied:
Where was the shop used?

Bronzal replied:
In original file
![Uploading 68DE90A7-6A74-4DE8-B7FC-B4C9FAEC2492.png…]()

chromaerror replied:
Ah yes. My mistake. Shop isn't unused, stats is.

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -33,22 +50,23 @@ new const Map_Interiors[] =
 
 const BLUE_COLOUR = 0x0000BBAA;
 const RED_COLOUR = 0xAA3333AA;
+new TEAM_BLUE_PLAYERS =0,TEAM_RED_PLAYERS=1;
```
Why not place these definitions on the same line always? Why is red initialized at 1 when it probably doesn't have 1 player at this point?

Bronzal replied:
It has different definition, just variable name isn’t correct I believe

chromaerror replied:
I mean, why would red team start at 1 player? Obviously if there's no players online then red team player amount probably shouldn't be 1 in tracking either?

Bronzal replied:
Let me explain.
Team blue and red is defined as 0 and 1
When player connects to server and blue is 0 it assigns them to blue team,
Now team blue is 2  and red is 1
Next player joins to server and now assigns to red team 
Blue 2 red 3
And loops.

chromaerror replied:
Why not initialize both at 0, then check if they both are 0, assign to one team (blue for example) +1 to that team, and then just check which one has more players? Wouldn't that be more logical and still preserve the balance?

Are you ever subtracting anything from the number, or just allow it to grow +2 every time a team is assigned? This can lead to teams being unbalanced after people have quit the game as the variables aren't being reduced from the team that the player left from. This can also (in theory) cause the variables to grow to ridiculous proportions, eat a whole lot of memory, and grow out of their size limits

Bronzal replied:
I’m assigning them to the default value when map changes and about subtracting. I believe I putted it in OnPlayerDisconnect.

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -135,7 +134,10 @@ public OnMapUpdate()
-                else SetPlayerTeam(i,_:RED_TEAM),TEAM_RED_PLAYERS+=2;
+                else
+                {
+                    SetPlayerTeam(i,_:RED_TEAM),TEAM_RED_PLAYERS+=2;
```
Why not do like
```pawn
SetPlayerTeam(i, _:RED_TEAM);
TEAM_RED_PLAYERS += 2;
```

Why are they on the same line, when they are different statements?

Bronzal replied:
Because I'm used to write like this.

chromaerror replied:
I'd still like to say that this is open-source, so it's not only what you are used to matter, but we should aim for clear code that includes best practices. This isn't that.

Also, i'd like to request for you to change how the teams are balanced as well. Use normal numbers instead that represent the amount of players, instead of doing some +2 magic

Bronzal replied:
The only easy thing that came up in my head rn is that to make Boolean like NextPlayerJoinsToBlueTeam.

chromaerror replied:
Why not just set them both to 0, then check if the both of them are zero when player joins, just decide to add that player to blue team, and then keep running the check which one is lower and add to the opposite team. I see no problem in that

Bronzal replied:
```c
if((BLUE_TEAM_PLAYERS - RED_TEAM_PLAYERS) == 0)
{
    SetPlayerTeam(playerid, _:BLUE_TEAM);
    BLUE_TEAM_PLAYERS++;
} 
else
{
    SetPlayerTeam(playerid, _:RED_TEAM);
    RED_TEAM_PLAYERS++;
}
```
I believe it's clear rn

Y-Less replied:
Except if `RED_TEAM_PLAYERS > BLUE_TEAM_PLAYERS` the subtraction will be negative, thus not `0`, so more players will get added to the red team, despite the fact that it is already the bigger team.

Bronzal replied:
![pic](https://user-images.githubusercontent.com/51256560/66845156-8a122800-ef3d-11e9-8dad-861c53dd9849.png)

Bronzal replied:
![image](https://user-images.githubusercontent.com/51256560/66845856-d3af4280-ef3e-11e9-807c-d65c46ab6064.png)
and also they assing correctly

Bronzal replied:
I'm checking them in OnPlayerDisconnect
![image](https://user-images.githubusercontent.com/51256560/66846158-5fc16a00-ef3f-11e9-8c7f-121216edf5c3.png)

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -130,20 +156,20 @@ public OnMapUpdate()
             if(IsPlayerConnected(i))
             {
-                if(TEAM_BLUE_PLAYERS <= TEAM_RED_PLAYERS)
+                if((BLUE_TEAM_PLAYERS - RED_TEAM_PLAYERS) == 0)
```
Whats going on here? Why not just check explicitly if `BLUE_TEAM_PLAYERS == 0 && RED_TEAM_PLAYERS == 0`? Then you can do the checks which team has more players and act based on that

Bronzal replied:
Because it is simple and works.

Bronzal replied:
easy thing is going here
if there is no difference between players it adds to blue team
if there is it add to red team

Y-Less replied:
Except it doesn't work.  Just use `<` like anyone else and use `++` instead of `+= 2`.

Bronzal replied:
in new version I have different assigner to teams and it works == works.

chromaerror commented on pawn/simpletdm-compat/simpletdm.pwn:
```diff
@@ -69,6 +71,30 @@ new
+        }
+        array[i] = teamnumber;
+    }
+    printf("RED TEAM PLAYERS %d | BLUE TEAM PLAYERS %d", RED_TEAM_PLAYERS,BLUE_TEAM_PLAYERS);
```
I have absolutely no idea whats going on between lines 74-97 and why. I mean I understand the code, but whats the function and why?

Bronzal replied:
This was for debug.

Bronzal replied:
And it's wrong version i re uploaded immediately

chromaerror replied:
Oh it was an old version. Sorry, then this is not relevant.

chromaerror commented:
@Bronzal Are you planning on fixing the comments?

Bronzal commented:
I need to fork it again :X
